File: app/lib/models/join_request.dart
import 'user.dart';
import 'ride.dart';

class JoinRequest {
  final String id;
  final String rideId;
  final String requesterId;
  final String status;
  final DateTime createdAt;
  final User? requester;
  final Ride? ride;
  final int unreadCount;
  final DateTime? lastReadOwner;
  final DateTime? lastReadRequester;

  JoinRequest({
    required this.id,
    required this.rideId,
    required this.requesterId,
    required this.status,
    required this.createdAt,
    this.requester,
    this.ride,
    this.unreadCount = 0,
    this.lastReadOwner,
    this.lastReadRequester,
  });

  factory JoinRequest.fromJson(Map<String, dynamic> json) {
    return JoinRequest(
      id: json['_id'] ?? json['id'],
      rideId: json['rideId'] is Map ? json['rideId']['_id'] : json['rideId'],
      requesterId: json['requesterId'] is Map
          ? json['requesterId']['_id']
          : json['requesterId'],
      status: json['status'],
      createdAt: DateTime.parse(json['createdAt']),
      requester: json['requester'] != null
          ? User.fromJson(json['requester'])
          : null,
      ride: json['rideId'] is Map ? Ride.fromJson(json['rideId']) : null,
      unreadCount: json['unreadCount'] ?? 0,
      lastReadOwner: json['lastReadOwner'] != null
          ? DateTime.parse(json['lastReadOwner'])
          : null,
      lastReadRequester: json['lastReadRequester'] != null
          ? DateTime.parse(json['lastReadRequester'])
          : null,
    );
  }

  bool get isPending => status == 'pending';
  bool get isAccepted => status == 'accepted';
  bool get isRejected => status == 'rejected';
  bool get hasUnread => unreadCount > 0;

  JoinRequest copyWith({
    String? status,
    DateTime? lastReadOwner,
    DateTime? lastReadRequester,
    int? unreadCount,
  }) {
    return JoinRequest(
      id: id,
      rideId: rideId,
      requesterId: requesterId,
      status: status ?? this.status,
      createdAt: createdAt,
      requester: requester,
      ride: ride,
      unreadCount: unreadCount ?? this.unreadCount,
      lastReadOwner: lastReadOwner ?? this.lastReadOwner,
      lastReadRequester: lastReadRequester ?? this.lastReadRequester,
    );
  }
}

File: backend/src/models/JoinRequest.js
const mongoose = require('mongoose');

const joinRequestSchema = new mongoose.Schema({
    rideId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Ride',
        required: true
    },
    requesterId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    status: {
        type: String,
        enum: ['pending', 'accepted', 'rejected'],
        default: 'pending'
    },
    // Track when each user last read messages
    lastReadOwner: {
        type: Date,
        default: Date.now
    },
    lastReadRequester: {
        type: Date,
        default: Date.now
    }
}, {
    timestamps: true
});

// Compound index to prevent duplicate join requests
joinRequestSchema.index({ rideId: 1, requesterId: 1 }, { unique: true });

// Index for querying requests by ride
joinRequestSchema.index({ rideId: 1, status: 1 });

// Virtual to populate requester details
joinRequestSchema.virtual('requester', {
    ref: 'User',
    localField: 'requesterId',
    foreignField: '_id',
    justOne: true
});

joinRequestSchema.set('toJSON', { virtuals: true });
joinRequestSchema.set('toObject', { virtuals: true });

module.exports = mongoose.model('JoinRequest', joinRequestSchema);
